examples/rbcd_attack_poisoning.py

```diff
@@ -0,0 +1,197 @@
+
+import torch_geometric.transforms as T
+from torch_geometric.datasets import Planetoid
+from torch_geometric.nn import GCNConv, PRBCDAttack
+from torch_geometric.nn.conv.gcn_conv import gcn_norm
+
+dataset = 'Cora'
+path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'Planetoid')
+# IMPORTANT: Edge weights are being ignored later and most adjacency
+# preprocessings should be part of the model (part of backpropagation)
+dataset = Planetoid(path, dataset, transform=T.NormalizeFeatures())
+data = dataset[0]
+
+
+class GCN(torch.nn.Module):
+    """GCN that normalizes adjacency matrix once for all layers (no cache)."""
+    def __init__(self, hidden_dim: int = 16):
+        super().__init__()
+        gcn_conv = functools.partial(
+            GCNConv,
+            cached=False,  # Important to backpropagate to the adj. matrix
+            add_self_loops=False,  # We add them in `self.gcn_norm`
+            normalize=False  # It is better to normalize once
+        )
+        self.conv1 = gcn_conv(dataset.num_features, hidden_dim)
+        self.conv2 = gcn_conv(hidden_dim, dataset.num_classes)
+        self.norm = gcn_norm
+
+    def forward(self, x, edge_index, edge_weight=None, skip_norm=False):
+        # Normalizing once lowers memory footprint and caching is not possible
+        # during attack (for training it would be possible)
+        if not skip_norm:
+            edge_index, edge_weight = self.norm(edge_index, edge_weight,
+                                                num_nodes=x.size(0),
+                                                add_self_loops=True)
+        x = F.relu(self.conv1(x, edge_index, edge_weight))
+        x = F.dropout(x, training=self.training)
+        x = self.conv2(x, edge_index, edge_weight)
+        return x
+
+    def reset_parameters(self):
+        self.conv1.reset_parameters()
+        self.conv2.reset_parameters()
+
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+gcn = GCN().to(device)
+data = data.to(device)
+
+x, edge_index, y = data.x, data.edge_index, data.y
+
+
+def train(model, x, edge_index, edge_weight=None, epochs=200, lr=0.01,
+          weight_decay=5e-4):
+    optimizer = torch.optim.Adam(model.parameters(), lr=lr,
+                                 weight_decay=weight_decay)
+    for _ in range(epochs):
+        optimizer.zero_grad()
+        prediction = model(x, edge_index, edge_weight)
+        loss = F.cross_entropy(prediction[data.train_mask],
+                               data.y[data.train_mask])
+        loss.backward()
+        optimizer.step()
+
+
+def accuracy(pred, lab, mask):
+    return (pred.argmax(-1)[mask] == lab[mask]).float().mean()
+
+
+gcn.train()
+train(gcn, x, edge_index)
+gcn.eval()
+
+print('\n------------- GCN: Global Poisoning -------------\n')
+
+clean_accuracy = accuracy(gcn(x, edge_index), y, data.test_mask).item()
+print(f'Clean accuracy: {clean_accuracy:.3f}')
+
+
+class PoisoningPRBCDAttack(PRBCDAttack):
```

Comment:
Any reason this is implemented in the example rather than as a module in `nn/models`?

Reply:
The main reason is that the train loop can have so many different variations. Of course, we could pass a method that defines the training procedure instead of tons of configurations. Moreover, the implementation here requires an additional dependency (i.e., `higher`).

If you prefer to have the "poisoning" as a dedicated model or configurable option, I am more than happy to implement it.